### PR TITLE
Add `Series.str`,`Series.dt`, and `Series.cat` accessors to docs

### LINKED ISSUE #8599: `Series.str` / `Series.dt` methods missing from Docs
Dask has many str accessor methods implemented but these are all missing from the documentation. For example, Dask supports `Series.str.capitalize` but it's missing in https://docs.dask.org/en/stable/dataframe-api.html.

xref: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.capitalize.html

### PR DESCRIPTION
**Pull request by scharlottej13**
- [ ] Closes #8599
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

I added the `Series.str` methods and `Series.dt` accessors to `dataframe-api.rst`, but am having trouble generating the autosummary. The specific Sphinx warning is `WARNING: autosummary: failed to import Series.str.zfill`, e.g., for all newly added lines.

I first tried to mimic the pandas docs, copying the changes made in `https://github.com/pandas-dev/pandas/pull/9322`. Then I tried using [`sphinx-autosummary-accessors`](https://github.com/xarray-contrib/sphinx-autosummary-accessors/tree/stable), which is reflected in this PR. Would very much appreciate ideas/suggestions at this point, I'm not quite sure what else to try!

### CODE REVIEW

**Comment by keewis** [comment]
`sphinx-autosummary-accessors` requires the attribute to be accessible from the class, so e.g. `Series.str.zfill` shouldn't raise an `AttributeError`.

The reason this currently fails is because [`Series.str`](https://github.com/dask/dask/blob/a27b437da2594fd5c15f85736aa520feb726ddbd/dask/dataframe/core.py#L3332-L3335) uses [`cache_readonly`](https://github.com/pandas-dev/pandas/blob/8e0baa2360a998f76a8e70dbee7bbf7339caf501/pandas/_libs/properties.pyx#L10-L42), which works like the builtin `property`: if accessed on the class object (`__get__` is called with `obj=None`), it will return itself instead of the object it wraps. See also pydata/xarray#3988.

In summary, you might need to use `CachedAccessor` instead of `cache_readonly` like [pandas](https://github.com/pandas-dev/pandas/blob/8e0baa2360a998f76a8e70dbee7bbf7339caf501/pandas/core/series.py#L5637-5641), get `pandas` to change their implementation of `cache_readonly`, or reimplement it here.

**Comment by scharlottej13** [comment]
Thanks @keewis for taking the time to explain this. After looking through the links you shared, it seems like even with using a custom descriptor (e.g. the [pandas example](https://github.com/pandas-dev/pandas/blob/8e0baa2360a998f76a8e70dbee7bbf7339caf501/pandas/core/series.py#L5637-5641) you linked) we still would run into problems, since in Dask the `Accessor` class [reads the methods from the `_meta` object](https://github.com/dask/dask/blob/a49f258b6fe9cd0b49521a734477a725eeb5478a/dask/dataframe/accessor.py#L80-L82).

After chatting w/ @ian-r-rose @ncclementi, creating a new [auto-documenter](https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx.application.Sphinx.add_autodocumenter) class might be an option, where I could get the docstrings from creating a `Series` instance. Is this something you tried already with `xarray`?

cc @jsignell

**Comment by keewis** [comment]
> since in Dask the Accessor class [reads the methods from the _meta object](https://github.com/dask/dask/blob/a49f258b6fe9cd0b49521a734477a725eeb5478a/dask/dataframe/accessor.py#L80-L82).

I see what you mean, this does indeed not work with the existing documenters in `sphinx-autosummary-accessors`.

> Is this something you tried already with `xarray`?

While I did consider to instantiate the class in the accessor documenter I never actually tried it because changing the descriptor seemed easier (and I could follow what `pandas` did).

In any case, feel free to ping me if you need help, and I would also be happy to include those instantiating `Accessor*` documenters and their templates in `sphinx-autosummary-accessors`.

**Comment by jcrist** [comment]
I've pushed a patch up that refactors our accessor implementation to define all methods/properties statically (and did a bunch of other cleanliness fixes while I was at it). Things are *almost* working now, but I'm a bit confused what's still broken. Current status:

- Code tests pass locally (so implementation works as intended)
- Docs are accessible on instances (e.g. `some_series.str.capitalize.__doc__` exists and is correct)
- Docs are accessible on class methods (e.g. `dd.Series.str.capitalize.__doc__` exists and is correct)
- Docs are accessible on class properties (e.g. `dd.Series.dt.date.__doc__` exists and is correct)
- `make html` runs (with some warnings) and generates docs pages for all accessor methods/properties

However, the accessor methods aren't documenting correctly. The accessor property docs are all working perfectly, but the methods show up with no docstring and as data.

image

I'm not sure what the issue is - @keewis do you have any thoughts?

**Comment by scharlottej13** [comment]
> However, the accessor methods aren't documenting correctly. The accessor property docs are all working perfectly, but the methods show up with no docstring and as data.
> 
> I'm not sure what the issue is - @keewis do you have any thoughts?

@jcrist this looks great!! I changed the template from `autosummary/accessor_attribute.rst` to `autosummary/accessor_method.rst` and I think this fixed the issue:

Screen Shot 2022-03-10 at 1 56 25 PM

File: docs/source/dataframe-api.rst [hunk 1]
```diff
@@ -260,6 +257,127 @@ Series
...
+~~~~~~~~~
+Similar to pandas, Dask provides dtype-specific methods under various accessors.
+These are separate namespaces within :class:`Series` that only apply to specific data types.
+
+Datetime Accessor
+*****************
+
+.. autosummary::
+   :toctree: generated/
+   :template: autosummary/accessor_attribute.rst
```
Comment (keewis) [inline docs/source/dataframe-api.rst:270]: you might have to split this in two parts: one for properties like `Series.dt.day` (using the `accessor_attribute` template), and another for the methods (using the `accessor_method` template).

Comment (jcrist) [reply docs/source/dataframe-api.rst]: Ah, nice catch. That's a bit unfortunate that this is needed. Knowing nothing about how sphinx works - is this a sphinx limitation, or is it something that this extension could handle with the same template for both?

**Comment by jcrist** [comment]
This should be good to go, will merge once tests pass.

File: docs/source/dataframe-api.rst (hunk 1 above)
Comment (keewis) [reply docs/source/dataframe-api.rst]: it's more like that is what `pandas` did when I last looked. It would probably possible to detect whether the attribute is a descriptor or a method and automatically use the appropriate documenter, but I never really tried.

Also, at least for the projects I work on the split serves as a reminder that those are called differently (I'm not too familiar with `dask.dataframe`, though)

**Comment by keewis** [comment]
the missing docstring for `Series.dt.freq` will have to be fixed in `pandas`, which also doesn't seem to have a [docstring](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.freq.html#pandas.Series.dt.freq) (although it has a [description](https://pandas.pydata.org/docs/reference/series.html#datetime-properties), which is a bit weird)
